src/bridges/tokemak/DepositBridge.sol

```diff
@@ -0,0 +1,337 @@
+
+    // Check whether the call is for withdrawal or deposit
+    bool isWithdrawal = _isWithdrawal(inputAssetA.erc20Address);
+    address tAsset = isWithdrawal ? inputAssetA.erc20Address : _getPoolAddress(inputAssetA.erc20Address);
+   
+    require(tAsset != address(0),"Tokemak DepositBridge: INVALID_INPUT");
+
+    // Withdraw or Deposit 
+    (outputValueA,isAsync) = isWithdrawal ? addWithdrawalNonce(interactionNonce, tAsset, totalInputValue) : _deposit(tAsset, totalInputValue, inputAssetA.erc20Address);
+
+    finalisePendingInteractions(MIN_GAS_FOR_FUNCTION_COMPLETION);
+
+  }
+
+  function _canWithdraw(address tAsset, uint256 inputValue) private returns (bool){
+       Ttoken tToken = Ttoken(tAsset);
+    
+    // Get our current request withdrawal data
+    (uint256 minCycle, uint256 requestedWithdrawalAmount) = tToken.requestedWithdrawals(address(this));
+
+    // Get current cycle index
+    uint256 currentCycleIndex = IManager(MANAGER).getCurrentCycleIndex();
+
+    // Check if need to request for withdraw first
+    if(inputValue > requestedWithdrawalAmount){
+      return false;
+    }
+
+    //Check if the withdrawal request is complete
+    if(currentCycleIndex < minCycle){
+      return false;
+    }
+    return true;
+  }
+
+  function addWithdrawalNonce(uint256 nonce, address tAsset, uint256 inputValue) private returns (uint256,bool ){
+    Ttoken tToken = Ttoken(tAsset);
+             tToken.requestWithdrawal(inputValue);
+
+    nonces.push(nonce);
+    pendingInteractions[nonce] =  Interaction(inputValue,tAsset);
+    return (0, true);
+  }
+
+  function _finaliseWithdraw(address tAsset, uint256 inputValue, uint256 nonce) private returns (uint256 outputValue,bool isAsync) {
+    Ttoken tToken = Ttoken(tAsset);
+
+    if(!_canWithdraw(tAsset,inputValue)){
+        isAsync = true;
+        outputValue = 0;
+        return (outputValue, isAsync);
+    }
+    // Approve Tokemak Pool to transfer tAsset
+    tToken.approve(tAsset, inputValue);
+
+    //Get asset address from tAsset
+    address asset = _getAssetAddress(tAsset);
+    IERC20 assetToken = IERC20(address(asset));
+
+    // Asset balance before withdrawal for calculating outputValue
+    uint256 beforeBalance = assetToken.balanceOf(address(this));
+
+    //Check if the pool is EthPool because withdrawal function is different
+    if(asset == WETH){
+      tToken.withdraw(inputValue, false);
+    }else{
+      tToken.withdraw(inputValue);
+    }
+
+    // Asset balance after withdrawal for calculating outputValue
+    uint256 afterBalance = assetToken.balanceOf(address(this));
+
+    outputValue = afterBalance - beforeBalance;
+
+    //Approve Rollup Processor to withdraw asset
+    assetToken.approve(rollupProcessor, outputValue);
+
+    delete pendingInteractions[nonce];
+    for(uint i = 0;i < nonces.length;i++){
+      if(nonces[i] == nonce){
+        delete nonces[i];
```

Comment:
Using delete here won't remove the element form the array, it will just set it to 0. So the array will continue to grow in size and your other loop where you iterate this will get increasingly expensive. Does the order of the array matter? If not then change `delete nonces[i]` to 

```
nonces[i] = nonces[nonces.length - 1];
nonces.pop();
```

Reply:
Order matters, because I think we want to first execute the transaction in starting of the queue